sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1055 @@
+		ocspHosts = conf.OcspServersOverride
+	}
+
+	var wg sync.WaitGroup
+
+	ocspStatuses := make([]*ocspStatus, len(ocspHosts))
+	ocspResponses := make([]*ocsp.Response, len(ocspHosts))
+	errors := make([]error, len(ocspHosts))
+
+	for i, ocspHost := range ocspHosts {
+		u, err := url.Parse(ocspHost)
+		if err != nil {
+			return nil, err
+		}
+
+		hostname := u.Hostname()
+
+		headers := make(map[string]string)
+		headers[httpHeaderContentType] = ocspRequestContentType
+		headers[httpHeaderAccept] = ocspResponseContentType
+		headers[httpHeaderContentLength] = strconv.Itoa(len(ocspReq))
+		headers[httpHeaderHost] = hostname
+		timeout := defaultOCSPResponderTimeout
+
+		ocspClient := retryablehttp.NewClient()
+		ocspClient.HTTPClient.Timeout = timeout
+		ocspClient.HTTPClient.Transport = newInsecureOcspTransport(conf.ExtraCas)
+
+		doRequest := func() error {
+			if conf.QueryAllServers {
+				defer wg.Done()
+			}
+			ocspRes, _, ocspS, err := c.retryOCSP(
+				ctx, ocspClient, retryablehttp.NewRequest, u, headers, ocspReq, issuer)
+			ocspResponses[i] = ocspRes
+			if err != nil {
+				errors[i] = err
+				return err
+			}
+			if ocspS.code != ocspSuccess {
+				ocspStatuses[i] = ocspS
+				return nil
+			}
+
+			ret, err := validateOCSP(ocspRes)
+			if err != nil {
+				errors[i] = err
+				return err
+			}
+			if isValidOCSPStatus(ret.code) {
+				ocspStatuses[i] = ret
+			}
+			return nil
+		}
+		if conf.QueryAllServers {
+			wg.Add(1)
+			go doRequest()
+		} else {
+			err = doRequest()
+			if err == nil {
+				break
+			}
+		}
+	}
+	if conf.QueryAllServers {
+		wg.Wait()
+	}
+	// Good by default
+	var ret *ocspStatus = &ocspStatus{code: ocspStatusGood}
+	ocspRes := ocspResponses[0]
+	for i, _ := range ocspHosts {
+		if errors[i] != nil {
+			return nil, errors[i]
+		} else if ocspStatuses[i] != nil && (!conf.QueryAllServers || ocspStatuses[i].code != ocspStatusGood) {
+			ret = ocspStatuses[i]
+			ocspRes = ocspResponses[i]
+			break
+		}
+	}
+
+	if !isValidOCSPStatus(ret.code) {
+		return ret, nil
+	}
+	v := ocspCachedResponse{
+		time:       float64(time.Now().UTC().Unix()),
+		producedAt: float64(ocspRes.ProducedAt.UTC().Unix()),
+		thisUpdate: float64(ocspRes.ThisUpdate.UTC().Unix()),
+		nextUpdate: float64(ocspRes.NextUpdate.UTC().Unix()),
+	}
+
+	c.ocspResponseCacheLock.Lock()
+	c.ocspResponseCache.Add(*encodedCertID, &v)
+	c.cacheUpdated = true
+	c.ocspResponseCacheLock.Unlock()
+	return ret, nil
+}
+
+func isValidOCSPStatus(status ocspStatusCode) bool {
+	return status == ocspStatusGood || status == ocspStatusRevoked || status == ocspStatusUnknown
+}
+
+type VerifyConfig struct {
+	ExtraCas            []*x509.Certificate
+	OcspServersOverride []string
+	OcspFailureMode     FailOpenMode
+	QueryAllServers     bool
+}
+
+// VerifyPeerCertificate verifies all of certificate revocation status
+func (c *Client) VerifyPeerCertificate(ctx context.Context, verifiedChains [][]*x509.Certificate, conf *VerifyConfig) error {
```

Comment:
As I think I mentioned on another thread, I'd perhaps suggest that we only validate the leaf cert by default and have an option (perhaps added at a later time) to do full chain validation. This'd also improve validation times by default... :)

Reply:
Updated to add a ValidateLeafCert exported function